File: src/ados/services/ui/pages/channel_hops.py
"""Channel Hops page — LCD widget for the HopSupervisor history.

Mirrors what ChannelHistoryChart.tsx renders on the GCS but as a PIL
painting on the 480x244 LCD canvas. The hop supervisor (drone profile,
inside ados-wfb) persists its snapshot to /run/ados/hop-supervisor.json
every 5 s; this page reads the file at 1 Hz and renders the recent
hop history as a step-after line + scatter markers colored by trigger
+ outcome.

Visual layout (top to bottom):
1. Header strip (24 px) — title + band + hop count
2. Chart region (~170 px) — X axis time, Y axis channel, step line
   through hops, scatter marker per hop, dashed reference line at the
   current channel
3. Legend strip (30 px) — colored dots for periodic / reactive /
   failed, plus the last-hop summary

Color semantics match the GCS chart so the operator sees the same
encoding on either surface:
- Green (palette.status_success / #22c55e): periodic + ok
- Amber (palette.status_warning / #f59e0b): reactive + ok
- Red   (palette.status_error   / #ef4444): any failed
- Blue  (palette.accent_primary / #3a82ff): current-channel ref line
"""

from __future__ import annotations

import json
import time
from typing import Any, ClassVar

from PIL import Image, ImageDraw

from ados.core.paths import HOP_SUPERVISOR_JSON
from ados.services.ui.dashboards.components import primitives as p
from ados.services.ui.pages.base import HitZone, PageContext
from ados.services.ui.touch.events import TouchGesture

PAGE_W = 480
PAGE_H = 244

# Vertical layout (numbers add up to PAGE_H exactly).
_HEADER_H = 24
_LEGEND_H = 30
_CHART_H = PAGE_H - _HEADER_H - _LEGEND_H  # 190 px

# 1 Hz file read; matches the supervisor's 5 s persist cadence with
# room to spare. Re-reading a small JSON every second is cheap.
_REFRESH_INTERVAL_S = 1.0

# Chart inner margins.
_CHART_LEFT_PAD = 36
_CHART_RIGHT_PAD = 12
_CHART_TOP_PAD = 8
_CHART_BOTTOM_PAD = 22  # leaves room for X-axis ticks

# Y-axis: a small breath above/below the actual data extrema so dots
# don't sit on the axis.
_Y_PAD_CHANNELS = 4

# Maximum hops the supervisor's snapshot caps at; we mirror it here
# to bound the loops.
_MAX_HOPS = 32


def _safe_dict(value: Any) -> dict:
    return value if isinstance(value, dict) else {}


class ChannelHopsPage:
    """LCD page showing the recent channel-hopping history.

    Pure read-only watch surface: no tap drilldowns, no actions.
    State lives in the JSON file written by HopSupervisor — the page
    re-fetches on every refresh and does not maintain its own ring
    buffer.
    """

    id: ClassVar[str] = "channel_hops"
    refresh_hz: ClassVar[float] = 2.0

    def __init__(self) -> None:
        self._hopping: dict[str, Any] = {}
        self._radio_channel: int | None = None
        self._last_refresh_at: float = 0.0

    # ── lifecycle ──────────────────────────────────────────────

    async def on_enter(self, ctx: PageContext) -> None:
        ctx.logger.info("channel_hops_enter")
        await self._refresh(ctx, force=True)

    async def on_leave(self, ctx: PageContext) -> None:
        ctx.logger.info("channel_hops_leave")

    # ── refresh ────────────────────────────────────────────────

    async def _refresh(self, ctx: PageContext, *, force: bool = False) -> None:
        now = time.monotonic()
        if not force and (now - self._last_refresh_at) < _REFRESH_INTERVAL_S:
            return
        self._last_refresh_at = now

        blob = self._read_run_json(str(HOP_SUPERVISOR_JSON))
        if blob:
            self._hopping = blob

        # The current radio channel pairs with the chart's reference
        # line. Try /api/wfb (same shape link_stats.py uses), fall
        # back to the state.link block populated by the OLED service's
        # internal 1 Hz poll. Either gives us the live channel; a
        # stale one is fine for a passive display.
        ch: int | None = None
        if ctx.http is not None:
            try:
                r = await ctx.http.get("/api/wfb", timeout=1.0)
                if r.status_code == 200:
                    body = r.json()
                    if isinstance(body, dict):
                        v = body.get("channel")
                        if isinstance(v, int) and v > 0:
                            ch = v
            except Exception:  # noqa: BLE001
                pass
        if ch is None:
            link = ctx.state.get("link") if hasattr(ctx, "state") else None
            if isinstance(link, dict):
                v = link.get("channel")
                if isinstance(v, int) and v > 0:
                    ch = v
        self._radio_channel = ch

    @staticmethod
    def _read_run_json(path: str) -> dict:
        try:
            with open(path) as f:
                blob = json.load(f)
                return blob if isinstance(blob, dict) else {}
        except (FileNotFoundError, ValueError, OSError):
            return {}

    # ── render ─────────────────────────────────────────────────

    async def render(self, ctx: PageContext) -> Image.Image:
        await self._refresh(ctx)
        palette = ctx.palette
        img = Image.new("RGB", (PAGE_W, PAGE_H), palette.bg_primary)

        history = self._history()

        # Header strip is the same in both empty + populated states.
        self._draw_header(img, palette, history)

        if not history:
            self._draw_empty(img, palette)
            return img

        # Chart region: axes + reference line + step-after series + dots.
        self._draw_chart(img, palette, history)
        self._draw_legend(img, palette, history)
        return img

    # ── data accessors ──────────────────────────────────────────

    def _history(self) -> list[dict[str, Any]]:
        raw = self._hopping.get("history") if self._hopping else None
        if not isinstance(raw, list):
            return []
        out: list[dict[str, Any]] = []
        for entry in raw[-_MAX_HOPS:]:
            if not isinstance(entry, dict):
                continue
            if not all(k in entry for k in ("at", "from", "to", "ok")):
                continue
            out.append(entry)
        return out

    def _band(self) -> str:
        v = self._hopping.get("band") if self._hopping else None
        return str(v) if isinstance(v, str) and v else "—"

    # ── band drawers ───────────────────────────────────────────

    def _draw_header(self, img, palette, history: list[dict[str, Any]]) -> None:
        d = ImageDraw.Draw(img)
        d.rectangle((0, 0, PAGE_W - 1, _HEADER_H - 1), fill=palette.bg_secondary)
        d.line(
            (0, _HEADER_H - 1, PAGE_W - 1, _HEADER_H - 1),
            fill=palette.border_default,
        )
        title_f = p.font("sans_bold", 11)
        d.text((12, 6), "CHANNEL HOPS", fill=palette.text_tertiary, font=title_f)
        right = f"{self._band()} · {len(history)} hops"
        # Right-justify approximately; small font so width is small.
        d.text((PAGE_W - 150, 6), right, fill=palette.text_secondary, font=title_f)

    def _draw_empty(self, img, palette) -> None:
        d = ImageDraw.Draw(img)
        big_f = p.font("sans_bold", 14)
        small_f = p.font("sans_regular", 11)
        msg = "No hops yet"
        sub = (
            f"current channel {self._radio_channel}"
            if self._radio_channel is not None
            else "supervisor is armed"
        )
        # Roughly center vertically inside the chart band.
        d.text(
            (PAGE_W // 2 - 50, _HEADER_H + _CHART_H // 2 - 16),
            msg,
            fill=palette.text_primary,
            font=big_f,
        )
        d.text(
            (PAGE_W // 2 - 80, _HEADER_H + _CHART_H // 2 + 4),
            sub,
            fill=palette.text_tertiary,
            font=small_f,
        )

    def _draw_chart(
        self, img, palette, history: list[dict[str, Any]],
    ) -> None:
        d = ImageDraw.Draw(img)
        chart_x0 = _CHART_LEFT_PAD
        chart_y0 = _HEADER_H + _CHART_TOP_PAD
        chart_x1 = PAGE_W - _CHART_RIGHT_PAD
        chart_y1 = _HEADER_H + _CHART_H - _CHART_BOTTOM_PAD
        chart_w = chart_x1 - chart_x0
        chart_h = chart_y1 - chart_y0

        # Y-axis domain: span the actual hop channels + the live ref
        # channel, with a small pad either side.
        ys = [int(e["to"]) for e in history]
        if self._radio_channel is not None:
            ys.append(self._radio_channel)
        y_min = max(1, min(ys) - _Y_PAD_CHANNELS)
        y_max = min(165, max(ys) + _Y_PAD_CHANNELS)
        if y_max <= y_min:
            y_max = y_min + 1

        # X-axis domain: seconds-from-start of the oldest entry.
        t0 = float(history[0]["at"])
        t_last = float(history[-1]["at"])
        x_span = max(1.0, t_last - t0)

        def to_px(at: float, ch: float) -> tuple[int, int]:
            tx = (at - t0) / x_span
            ty = (ch - y_min) / (y_max - y_min)
            px = chart_x0 + int(tx * chart_w)
            py = chart_y1 - int(ty * chart_h)
            return px, py

        # Background.
        d.rectangle(
            (chart_x0 - 1, chart_y0 - 1, chart_x1 + 1, chart_y1 + 1),
            outline=palette.border_default,
            fill=palette.bg_secondary,
        )

        # Y-axis tick labels at min, mid, max.
        axis_f = p.font("mono_regular", 9)
        for ch in (y_min, (y_min + y_max) // 2, y_max):
            _, py = to_px(t0, ch)
            d.text(
                (4, py - 6), str(ch), fill=palette.text_tertiary, font=axis_f,
            )
            d.line(
                (chart_x0 - 2, py, chart_x0, py),
                fill=palette.border_default,
            )

        # X-axis tick labels: oldest (left), midpoint, newest (right).
        now_s = time.time()
        for t_at, anchor_x in (
            (t0, chart_x0),
            ((t0 + t_last) / 2.0, (chart_x0 + chart_x1) // 2),
            (t_last, chart_x1 - 30),
        ):
            delta = int(now_s - t_at)
            label = "now" if delta <= 1 else f"-{delta}s"
            d.text(
                (anchor_x, chart_y1 + 4),
                label,
                fill=palette.text_tertiary,
                font=axis_f,
            )

        # Dashed reference line at the current channel (if known and
        # inside the visible range).
        if self._radio_channel is not None and y_min <= self._radio_channel <= y_max:
            _, ref_py = to_px(t0, float(self._radio_channel))
            self._draw_dashed_hline(
                d, chart_x0, ref_py, chart_x1, palette.accent_primary,
            )

        # Step-after path through (from -> to) pairs.
        # Build a series of (px, py) pairs by walking through history;
        # each hop emits a horizontal segment at the "from" channel
        # and a vertical jump to the "to" channel.
        line_color = palette.text_secondary
        prev_px: int | None = None
        prev_py: int | None = None
        for i, entry in enumerate(history):
            at = float(entry["at"])
            from_ch = float(entry["from"])
            to_ch = float(entry["to"])
            from_px, from_py = to_px(at, from_ch)
            _to_px_, to_py = to_px(at, to_ch)
            # Horizontal extension from the previous point to this hop's
            # x position at the previous channel.
            if prev_px is not None and prev_py is not None:
                d.line(
                    (prev_px, prev_py, from_px, prev_py),
                    fill=line_color,
                    width=1,
                )
                # Vertical from previous channel to this hop's "from"
                # channel (usually equal — this is a continuity check).
                if prev_py != from_py:
                    d.line(
                        (from_px, prev_py, from_px, from_py),
                        fill=line_color,
                        width=1,
                    )
            # Vertical jump at this hop to the new channel.
            d.line(
                (from_px, from_py, from_px, to_py),
                fill=line_color,
                width=1,
            )
            prev_px, prev_py = from_px, to_py

        # Extend the last step to the right edge so the operator sees
        # "we're on this channel right now."
        if prev_px is not None and prev_py is not None and prev_px < chart_x1:
            d.line(
                (prev_px, prev_py, chart_x1, prev_py),
                fill=line_color,
                width=1,
            )

        # Scatter markers per hop, colored by trigger + outcome.
        for entry in history:
            at = float(entry["at"])
            to_ch = float(entry["to"])
            ok = bool(entry.get("ok", False))
            trigger = str(entry.get("trigger", "periodic"))
            color = self._marker_color(palette, trigger, ok)
            cx, cy = to_px(at, to_ch)
            d.ellipse(
                (cx - 3, cy - 3, cx + 3, cy + 3),
                fill=color,
                outline=palette.bg_primary,
                width=1,
            )

    @staticmethod
    def _marker_color(palette, trigger: str, ok: bool):
        if not ok:
            return palette.status_error
        if trigger == "reactive":
            return palette.status_warning
        return palette.status_success

    @staticmethod
    def _draw_dashed_hline(d, x0: int, y: int, x1: int, color) -> None:
        seg = 4
        gap = 3
        x = x0
        while x < x1:
            d.line((x, y, min(x + seg, x1), y), fill=color, width=1)
            x += seg + gap

    def _draw_legend(
        self, img, palette, history: list[dict[str, Any]],
    ) -> None:
        d = ImageDraw.Draw(img)
        legend_y = _HEADER_H + _CHART_H
        d.rectangle(
            (0, legend_y, PAGE_W - 1, PAGE_H - 1),
            fill=palette.bg_secondary,
        )
        d.line((0, legend_y, PAGE_W - 1, legend_y), fill=palette.border_default)
        f = p.font("sans_regular", 10)

        # Left-side legend: three colored dots + labels.
        items = (
            (palette.status_success, "periodic"),
            (palette.status_warning, "reactive"),
            (palette.status_error, "failed"),
        )
        cx = 14
        cy = legend_y + 14
        for color, label in items:
            d.ellipse((cx - 4, cy - 4, cx + 4, cy + 4), fill=color)
            d.text((cx + 8, cy - 6), label, fill=palette.text_primary, font=f)
            cx += 90

        # Right-side: last-hop summary "last: -56s (40 -> 44)".
        last = history[-1]
        try:
            delta = int(time.time() - float(last["at"]))
        except (KeyError, TypeError, ValueError):
            delta = 0
        ago = f"-{delta}s"
        summary = (
            f"last: {ago} ({int(last['from'])} -> {int(last['to'])})"
        )
        d.text(
            (PAGE_W - 170, cy - 6),
            summary,
            fill=palette.text_secondary,
            font=f,
        )

    # ── input ──────────────────────────────────────────────────

    def hit_zones(self, ctx: PageContext) -> list[HitZone]:
        # Pure read-only surface — no tap drilldowns.
        return []

    async def on_touch(
        self,
        ctx: PageContext,
        zone: HitZone,
        gesture: TouchGesture,
    ) -> None:
        # No hit zones registered; nothing to dispatch.
        return None

File: crates/ados-display/src/pages/channel_hops.rs
//! Channel Hops page — the channel-hopping history chart.
//!
//! A top-level tab and a read-only watch surface. It renders the recent hop
//! history from [`PageContext::hopping`] as a step-after line plus scatter
//! markers colored by trigger (periodic green, reactive amber) and outcome
//! (failed red), with a dashed reference line at the current radio channel, a
//! header strip (title + band + hop count), and a legend strip (color key plus
//! the last-hop summary).
//!
//! The content region is a 480x244 frame that sits below the persistent top
//! status bar, so every page-local y derives from [`CONTENT_Y`]. The page is a
//! pure composer: no tap drilldowns, only the bottom-bar tab zones.

use std::time::{SystemTime, UNIX_EPOCH};

use embedded_graphics::pixelcolor::Rgb888;

use crate::graphics::fonts::{FontFace, LoadedFont};
use crate::graphics::palette::Palette;
use crate::graphics::primitives::{fill_circle, fill_rect, fill_rect_outline, line, text, Canvas};
use crate::pages::{
    blank_panel, HitZone, HopEntry, Page, PageContext, CONTENT_H, CONTENT_Y, PANEL_W,
};
use crate::widgets::{bottom_bar_zones, draw_bottom_bar, draw_top_bar};

/// Content-region width.
const PAGE_W: i32 = PANEL_W as i32;
/// Content-region height (the 480x244 frame below the top status bar).
const PAGE_H: i32 = CONTENT_H as i32;
/// Panel-global y offset of the content region's top edge.
const OY: i32 = CONTENT_Y as i32;

/// Header strip height.
const HEADER_H: i32 = 24;
/// Legend strip height.
const LEGEND_H: i32 = 30;
/// Chart band height (the remainder between header and legend).
const CHART_H: i32 = PAGE_H - HEADER_H - LEGEND_H;

/// Chart inner padding.
const CHART_LEFT_PAD: i32 = 36;
const CHART_RIGHT_PAD: i32 = 12;
const CHART_TOP_PAD: i32 = 8;
const CHART_BOTTOM_PAD: i32 = 22;

/// Y-axis breathing room above / below the channel extrema.
const Y_PAD_CHANNELS: i64 = 4;

/// Cap on how many of the most recent hops the chart walks.
const MAX_HOPS: usize = 32;

/// The channel-hopping history surface, registered as `channel_hops`.
pub struct ChannelHopsPage;

impl Page for ChannelHopsPage {
    fn id(&self) -> &'static str {
        "channel_hops"
    }

    fn refresh_hz(&self) -> f32 {
        2.0
    }

    fn render(&self, ctx: &PageContext, palette: &Palette) -> Canvas {
        let mut canvas = blank_panel(palette);
        draw_top_bar(
            &mut canvas,
            palette,
            &ctx.hostname,
            ctx.role.current.as_deref().unwrap_or("unset"),
            ctx.system.cpu_pct,
            ctx.system.ram_used_mb,
            ctx.system.ram_total_mb,
            ctx.system.temp_c,
            &ctx.clock,
        );
        // Ground the content region.
        fill_rect(
            &mut canvas,
            0,
            OY,
            PAGE_W - 1,
            OY + PAGE_H - 1,
            palette.bg_primary,
        );

        let history = recent_history(&ctx.hopping.history);
        let radio_channel = ctx.hopping.radio_channel;

        draw_header(&mut canvas, palette, &ctx.hopping.band, history.len());
        if history.is_empty() {
            draw_empty(&mut canvas, palette, radio_channel);
        } else {
            draw_chart(&mut canvas, palette, history, radio_channel);
            draw_legend(&mut canvas, palette, history);
        }

        draw_bottom_bar(&mut canvas, palette, self.id());
        canvas
    }

    fn hit_zones(&self, _ctx: &PageContext) -> Vec<HitZone> {
        bottom_bar_zones()
    }
}

/// The most recent hops, capped at [`MAX_HOPS`].
fn recent_history(history: &[HopEntry]) -> &[HopEntry] {
    let from = history.len().saturating_sub(MAX_HOPS);
    &history[from..]
}

/// Current wall-clock as seconds since the Unix epoch, for the "-Ns" labels.
fn now_seconds() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

/// The band label, or an em dash when the supervisor has not reported one.
fn band_label(band: &Option<String>) -> String {
    match band {
        Some(b) if !b.is_empty() => b.clone(),
        _ => "—".to_string(),
    }
}

/// Paint the header strip: a caps title on the left and a band + hop-count
/// summary on the right, over a divider.
fn draw_header(canvas: &mut Canvas, palette: &Palette, band: &Option<String>, hops: usize) {
    fill_rect(
        canvas,
        0,
        OY,
        PAGE_W - 1,
        OY + HEADER_H - 1,
        palette.bg_secondary,
    );
    line(
        canvas,
        0,
        OY + HEADER_H - 1,
        PAGE_W - 1,
        OY + HEADER_H - 1,
        palette.border_default,
    );
    let title_f = LoadedFont::new(FontFace::SansBold, 11);
    text(
        canvas,
        &title_f,
        "CHANNEL HOPS",
        12,
        OY + 6,
        palette.text_tertiary,
    );
    let right = format!("{} · {} hops", band_label(band), hops);
    text(
        canvas,
        &title_f,
        &right,
        PAGE_W - 150,
        OY + 6,
        palette.text_secondary,
    );
}

/// Paint the empty-state body: a "no hops yet" headline and a sub-line that
/// reports the current channel (or that the supervisor is armed).
fn draw_empty(canvas: &mut Canvas, palette: &Palette, radio_channel: Option<i64>) {
    let big_f = LoadedFont::new(FontFace::SansBold, 14);
    let small_f = LoadedFont::new(FontFace::SansRegular, 11);
    let msg = "No hops yet";
    let sub = match radio_channel {
        Some(ch) => format!("current channel {ch}"),
        None => "supervisor is armed".to_string(),
    };
    text(
        canvas,
        &big_f,
        msg,
        PAGE_W / 2 - 50,
        OY + HEADER_H + CHART_H / 2 - 16,
        palette.text_primary,
    );
    text(
        canvas,
        &small_f,
        &sub,
        PAGE_W / 2 - 80,
        OY + HEADER_H + CHART_H / 2 + 4,
        palette.text_tertiary,
    );
}

/// Paint the chart band: axes + tick labels + dashed reference line + the
/// step-after path through the hops + a scatter marker per hop.
fn draw_chart(
    canvas: &mut Canvas,
    palette: &Palette,
    history: &[HopEntry],
    radio_channel: Option<i64>,
) {
    let chart_x0 = CHART_LEFT_PAD;
    let chart_y0 = OY + HEADER_H + CHART_TOP_PAD;
    let chart_x1 = PAGE_W - CHART_RIGHT_PAD;
    let chart_y1 = OY + HEADER_H + CHART_H - CHART_BOTTOM_PAD;
    let chart_w = chart_x1 - chart_x0;
    let chart_h = chart_y1 - chart_y0;

    // Y-axis domain: the hop "to" channels plus the live reference channel.
    let mut ys: Vec<i64> = history.iter().map(|e| e.to_channel).collect();
    if let Some(ch) = radio_channel {
        ys.push(ch);
    }
    let raw_min = ys.iter().copied().min().unwrap_or(1);
    let raw_max = ys.iter().copied().max().unwrap_or(1);
    let y_min = (raw_min - Y_PAD_CHANNELS).max(1);
    let mut y_max = (raw_max + Y_PAD_CHANNELS).min(165);
    if y_max <= y_min {
        y_max = y_min + 1;
    }

    // X-axis domain: seconds from the oldest entry.
    let t0 = history[0].at;
    let t_last = history[history.len() - 1].at;
    let x_span = (t_last - t0).max(1.0);

    // A hop record whose timestamps run newest-first (or any out-of-order /
    // sparse history) would drive `tx` far outside `0..=1`. PIL silently clips
    // off-canvas coordinates; the panel rasterizer instead walks every pixel of
    // the segment, so an unclamped coordinate billions of pixels off-screen
    // would blow the line algorithm. Clamp the normalized fractions to the
    // drawable band so the chart degrades to an edge-pinned point exactly where
    // PIL would clip it, and never feeds the rasterizer a runaway endpoint.
    let to_px = |at: f64, ch: f64| -> (i32, i32) {
        let tx = ((at - t0) / x_span).clamp(0.0, 1.0);
        let ty = ((ch - y_min as f64) / (y_max - y_min) as f64).clamp(0.0, 1.0);
        let px = chart_x0 + (tx * chart_w as f64) as i32;
        let py = chart_y1 - (ty * chart_h as f64) as i32;
        (px, py)
    };

    // Chart background.
    fill_rect_outline(
        canvas,
        chart_x0 - 1,
        chart_y0 - 1,
        chart_x1 + 1,
        chart_y1 + 1,
        palette.bg_secondary,
        palette.border_default,
    );

    // Y-axis tick labels at min, mid, max.
    let axis_f = LoadedFont::new(FontFace::MonoRegular, 9);
    for ch in [y_min, (y_min + y_max) / 2, y_max] {
        let (_, py) = to_px(t0, ch as f64);
        text(
            canvas,
            &axis_f,
            &ch.to_string(),
            4,
            py - 6,
            palette.text_tertiary,
        );
        line(
            canvas,
            chart_x0 - 2,
            py,
            chart_x0,
            py,
            palette.border_default,
        );
    }

    // X-axis tick labels: oldest (left), midpoint, newest (right).
    let now_s = now_seconds();
    let ticks = [
        (t0, chart_x0),
        ((t0 + t_last) / 2.0, (chart_x0 + chart_x1) / 2),
        (t_last, chart_x1 - 30),
    ];
    for (t_at, anchor_x) in ticks {
        let delta = (now_s - t_at) as i64;
        let label = if delta <= 1 {
            "now".to_string()
        } else {
            format!("-{delta}s")
        };
        text(
            canvas,
            &axis_f,
            &label,
            anchor_x,
            chart_y1 + 4,
            palette.text_tertiary,
        );
    }

    // Dashed reference line at the current channel, when inside the visible
    // range.
    if let Some(ch) = radio_channel {
        if y_min <= ch && ch <= y_max {
            let (_, ref_py) = to_px(t0, ch as f64);
            draw_dashed_hline(canvas, chart_x0, ref_py, chart_x1, palette.accent_primary);
        }
    }

    // Step-after path: each hop emits a horizontal run at the previous channel
    // and a vertical jump to its new channel.
    let line_color = palette.text_secondary;
    let mut prev: Option<(i32, i32)> = None;
    for entry in history {
        let (from_px, from_py) = to_px(entry.at, entry.from_channel as f64);
        let (_, to_py) = to_px(entry.at, entry.to_channel as f64);
        if let Some((prev_px, prev_py)) = prev {
            line(canvas, prev_px, prev_py, from_px, prev_py, line_color);
            if prev_py != from_py {
                line(canvas, from_px, prev_py, from_px, from_py, line_color);
            }
        }
        line(canvas, from_px, from_py, from_px, to_py, line_color);
        prev = Some((from_px, to_py));
    }
    // Extend the last step to the right edge — "we are on this channel now".
    if let Some((prev_px, prev_py)) = prev {
        if prev_px < chart_x1 {
            line(canvas, prev_px, prev_py, chart_x1, prev_py, line_color);
        }
    }

    // Scatter markers per hop, colored by trigger + outcome.
    for entry in history {
        let trigger = entry.trigger.as_deref().unwrap_or("periodic");
        let color = marker_color(palette, trigger, entry.ok);
        let (cx, cy) = to_px(entry.at, entry.to_channel as f64);
        fill_circle(canvas, cx, cy, 3, color, Some(palette.bg_primary));
    }
}

/// The scatter-marker color: red for a failed hop, amber for a reactive hop,
/// green for a successful periodic hop.
fn marker_color(palette: &Palette, trigger: &str, ok: bool) -> Rgb888 {
    if !ok {
        palette.status_error
    } else if trigger == "reactive" {
        palette.status_warning
    } else {
        palette.status_success
    }
}

/// Stroke a dashed horizontal line at `y` from `x0` to `x1`.
fn draw_dashed_hline(canvas: &mut Canvas, x0: i32, y: i32, x1: i32, color: Rgb888) {
    let seg = 4;
    let gap = 3;
    let mut x = x0;
    while x < x1 {
        line(canvas, x, y, (x + seg).min(x1), y, color);
        x += seg + gap;
    }
}

/// Paint the legend strip: three colored dots with labels on the left and the
/// last-hop summary on the right.
fn draw_legend(canvas: &mut Canvas, palette: &Palette, history: &[HopEntry]) {
    let legend_y = OY + HEADER_H + CHART_H;
    fill_rect(
        canvas,
        0,
        legend_y,
        PAGE_W - 1,
        OY + PAGE_H - 1,
        palette.bg_secondary,
    );
    line(
        canvas,
        0,
        legend_y,
        PAGE_W - 1,
        legend_y,
        palette.border_default,
    );
    let f = LoadedFont::new(FontFace::SansRegular, 10);

    let items = [
        (palette.status_success, "periodic"),
        (palette.status_warning, "reactive"),
        (palette.status_error, "failed"),
    ];
    let mut cx = 14;
    let cy = legend_y + 14;
    for (color, label) in items {
        fill_circle(canvas, cx, cy, 4, color, None);
        text(canvas, &f, label, cx + 8, cy - 6, palette.text_primary);
        cx += 90;
    }

    // Right side: "last: -Ns (from -> to)".
    let last = &history[history.len() - 1];
    let delta = (now_seconds() - last.at) as i64;
    let summary = format!(
        "last: -{}s ({} -> {})",
        delta.max(0),
        last.from_channel,
        last.to_channel
    );
    text(
        canvas,
        &f,
        &summary,
        PAGE_W - 170,
        cy - 6,
        palette.text_secondary,
    );
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::graphics::palette::DARK;

    fn hop(at: f64, from: i64, to: i64, ok: bool, trigger: &str) -> HopEntry {
        HopEntry {
            at,
            from_channel: from,
            to_channel: to,
            ok,
            trigger: Some(trigger.to_string()),
        }
    }

    #[test]
    fn empty_history_renders_with_tab_zones() {
        let page = ChannelHopsPage;
        let ctx = PageContext::default();
        let c = page.render(&ctx, &DARK);
        assert_eq!(c.width(), PANEL_W);
        assert_eq!(page.hit_zones(&ctx).len(), 5);
        assert_eq!(page.id(), "channel_hops");
    }

    #[test]
    fn populated_history_renders_chart_and_legend() {
        let page = ChannelHopsPage;
        let mut ctx = PageContext::default();
        ctx.hopping.band = Some("u-nii-3".to_string());
        ctx.hopping.radio_channel = Some(149);
        let base = 1_700_000_000.0;
        ctx.hopping.history = vec![
            hop(base, 149, 153, true, "periodic"),
            hop(base + 30.0, 153, 157, true, "reactive"),
            hop(base + 60.0, 157, 149, false, "reactive"),
        ];
        let c = page.render(&ctx, &DARK);
        assert_eq!(c.width(), PANEL_W);
    }

    #[test]
    fn history_is_capped_at_max_hops() {
        let base = 1_700_000_000.0;
        let many: Vec<HopEntry> = (0..50)
            .map(|i| hop(base + i as f64, 149, 153, true, "periodic"))
            .collect();
        let recent = recent_history(&many);
        assert_eq!(recent.len(), MAX_HOPS);
        // The tail is kept: the last entry survives the cap.
        assert_eq!(recent[recent.len() - 1].at, base + 49.0);
    }

    #[test]
    fn reversed_history_renders_without_overflow() {
        // A history whose timestamps run newest-first drives the time fraction
        // far outside 0..=1; the chart must clamp to the drawable band rather
        // than feed the line rasterizer a runaway endpoint.
        let page = ChannelHopsPage;
        let mut ctx = PageContext::default();
        ctx.hopping.radio_channel = Some(149);
        let base = 1_700_000_000.0;
        ctx.hopping.history = vec![
            hop(base + 120.0, 157, 149, true, "reactive"),
            hop(base + 60.0, 153, 157, true, "periodic"),
            hop(base, 149, 153, false, "reactive"),
        ];
        let c = page.render(&ctx, &DARK);
        assert_eq!(c.width(), PANEL_W);
        assert_eq!(c.height(), crate::pages::PANEL_H);
    }

    #[test]
    fn marker_color_follows_trigger_and_outcome() {
        assert_eq!(marker_color(&DARK, "periodic", false), DARK.status_error);
        assert_eq!(marker_color(&DARK, "reactive", true), DARK.status_warning);
        assert_eq!(marker_color(&DARK, "periodic", true), DARK.status_success);
    }
}
